src/promptflow/promptflow/tracing/_trace.py

```diff
@@ -120,60 +131,70 @@ def enrich_span_with_input(span, input):
 
 def enrich_span_with_trace_type(span, inputs, output, trace_type):
     if trace_type == TraceType.LLM:
+        # Handle the non-streaming output of LLM, the streaming output will be handled in traced_generator.
         token_collector.collect_openai_tokens(span, output)
-        enrich_span_with_llm_model(span, output)
+        enrich_span_with_llm_output(span, output)
     elif trace_type == TraceType.EMBEDDING:
         token_collector.collect_openai_tokens(span, output)
         enrich_span_with_embedding(span, inputs, output)
     enrich_span_with_openai_tokens(span, trace_type)
-    return enrich_span_with_output(span, output)
+    enrich_span_with_output(span, output)
+    # If the output is a generator, while the span is a valid span, we will trace the generator.
+    if isinstance(output, Iterator) and not isinstance(span, NonRecordingSpan):
+        output = traced_generator(span, inputs, output)
+    return output
 
 
-def traced_generator(generator, original_span: ReadableSpan):
+def traced_generator(original_span: ReadableSpan, inputs, generator):
     context = original_span.get_span_context()
     link = Link(context)
     # If start_trace is not called, the name of the original_span will be empty.
     with open_telemetry_tracer.start_as_current_span(
         f"Iterated({original_span.name})",
         links=[link],
     ) as span:
-        span.set_attributes(original_span.attributes)
+        enrich_span_with_original_attributes(span, original_span.attributes)
         generator_proxy = GeneratorProxy(generator)
         yield from generator_proxy
         generator_output = generator_proxy.items
 
-        # Enrich LLM span for openai steaming message
-        # TODO: Enrich LLM token count for streaming scenario
+        # Enrich LLM span for OpenAI streaming output
         if original_span.attributes["span_type"] == "LLM" and not IS_LEGACY_OPENAI:
             from openai.types.chat.chat_completion_chunk import ChatCompletionChunk
-            chunks = []
-            role = "assistant"
-            for item in generator_output:
-                if not isinstance(item, ChatCompletionChunk):
-                    continue
-                if item.choices and item.choices[0].delta.content:
-                    chunks.append(item.choices[0].delta.content)
-                    role = item.choices[0].delta.role or role
-            if chunks:
-                text = "".join(chunks)
-                message = {"content": text, "role": role}
-                span.set_attribute("llm.generated_message", serialize_attribute(message))
-        serialized_output = serialize_attribute(generator_output)
-        span.set_attribute("output", serialized_output)
+            from openai.types.completion import Completion
+
+            if generator_output and isinstance(generator_output[0], (ChatCompletionChunk, Completion)):
+                parser = OpenAIResponseParser(generator_output)
```

Comment:
(Optional) Let's separate OpenAIResponseParser into two, one for Chat, and another for Completion. They can inherit from the same abstract parent class.

Reply:
Done.